### ISSUE TITLE
Serve a dynamic sitemap at `/sitemap.xml`

### ISSUE THREAD

**Issue by timolegros** [hicommonwealth/commonwealth#6173]
## Description

Google currently can't crawl our site at all because of #6165 but even if we fix that issue crawlers may have a hard time on our site because we don't provide a sitemap that tells the crawlers where they should go. In other words, we don't tell Google which pages are important and should be indexed. Instead, pages are only indexed if they are linked to outside of Commonwealth.

We can resolve this issue by dynamically creating a sitemap. That means that every time a crawler requests `/sitemap.xml` we build an updated sitemap (could return a cached version that is periodically refreshed as well) with links to all major communities, posts, and pages. We can do this by constructing the URLS manually and then using a package like [sitemap][1].

## Engineering Requirements

- Create a new `/sitemap.xml` route that returns a relatively fresh sitemap
- Create a function which creates a sitemap using live data from the database (community + post URLs)

## Additional context

Providing a sitemap helps with prerendering as well since prerender.io can crawl our pages and prerender them periodically.

[1]: https://www.npmjs.com/package/sitemap

**Comment by jnaviask** [comment]
#4985 may be a helpful basis for generating the sitemap.

**Comment by lzach83** [comment]
Handing this back to @timolegros, we had some in depth conversations about how to solve this. It does not seem possible on the front-end due to the dynamic community / thread / discussions routes and one of the reqs being we want to include those.

**Comment by timolegros** [comment]
Since client-side generation didn't work we need to generate the sitemap on the backend. Scope of work is as follows:

- Figure out URL format for threads, general pages, and profiles
- Create script which paginates through all existing threads + profiles from the DB and create the links + append to sitemap
- Figure out how we save the sitemap on Heroku i.e. can this be a cron job or does it have to execute as part of release
- Create the sitemap route

**Comment by jnaviask** [comment]
Backlogging as lower priority concern.

**Comment by CowMuon** [comment]
Given that @timolegros is heroing notifications, this seems eminently handoffable to @burtonator.

**Comment by burtonator** [comment]
One of the main issues is multiple sitemaps as we're going to have a lot of pages. I'm going to check into that first and then work out a proposal.

I think I have a workaround to split everything up easily and use prerender to make this all simple.

**Comment by burtonator** [comment]
The big issue I think is going to be pagination of our threads. 

Our IDs are autogenerating so we can't use hash prefixes.

We can't use offset + length because that's too inefficient.  

I think what we could / should do is lge, gte ranges and then keep them limited by size.

Either that or just a "select *" and an order by clause that dumps everything out into a S3 bucket and have a fixed number of links per bucket.  Then use those bucket URLs in the sitemap.xml file

**Comment by burtonator** [comment]
For now I think we should probably break each unit into 50k links (which is the per sitemap limit).  

Ideally these files would be semi-immutable so that Google Maps doesn't constantly have to refetch them.

**Comment by burtonator** [comment]
We also have to setup the google search dashboard to make sure all these things actually work!

**Comment by burtonator** [comment]
Update.  I have the database scanning code done and the code to generate the sitemap and write it to S3 is done too.  I have to do some more hacking on the tests so I can get access to real data to test it with.  

Then I have to update our robots.txt to link to our dynamic sitemap and we should be good to go.  Probably another day worth of work.

**Comment by burtonator** [comment]
The threads pagination works now and uses the correct URL.  That was difficult as I had to reverse engineer how getThreadUrl works. Next I have to do profiles.

**Comment by burtonator** [comment]
OK. I think all the heavy lifting is done now.  The profiles and threads work. I also made it easy so we can export more URL types. 

Next steps are:

- improve the testing
- move it into a module with 'strict' types. (I already found that it missed a but with a nullable type)
- then find a machine in the cloud to turn it on in production
- update our robots.txt to link to the sitemap index file (easy)
